blog/_posts/2024-08-19-KAITO-finetuning.md

```diff
@@ -0,0 +1,104 @@
+---
+title: "Fine tune language models with KAITO on AKS"
+description: "Make your AI model even smarter, in just a few steps"
+date: 2024-08-19
+author: Sachi Desai
+categories: AI, Machine Learning, open-source, operator
+---
+
+**Introduction**
+
+You may have heard of the [Kubernetes AI Toolchain Operator](https://github.com/Azure/kaito/tree/main) (KAITO) announced at Ignite 2023 and KubeCon EU this year. The open source project has gained popularity in recent months by introducing a streamlined approach to AI model deployment and flexible infrastructure provisioning on Kubernetes. 
+
+With the [v0.3.0 release](https://github.com/Azure/kaito/releases/tag/v0.3.0) of KAITO, we’re excited to share that we’ve expanded the supported model library to include a preset configuration for Phi-3, but the biggest (any most exciting) addition is the ability to fine-tune open-source models. Fine-tuning is just one way of giving your foundation model additional training for a specific task to enhance accuracy, which ultimately improves the interaction with end-users.
+
+All you have to do is specify your model preset and tuning method, data set, and container registry details - KAITO will automate the rest and provide a lightweight output to pull into new inferencing jobs!
+
+**Fine-tune a Phi-3 model with KAITO**
+
+Let’s look at what fine-tuning looks like with KAITO. We’ll use Phi-3-mini for our sample deployment. The `Phi-3-mini-128K` model is a robust and high-performing option that’s also a convenient size for fine-tuning, so we’ll use that as an example. With KAITO v0.3.0 or higher, there is an additional tuning property added to the `Workspace` specification. Within the tuning property, you can specify additional information for the tuning job such as a preset configuration for the base model, fine-tuning method using LoRA or QLoRA, input datasets, and output location for the adapter.
+
+Here’s an example of what the fine-tuning workspace looks like when training on a dataset found on HuggingFace:
+
+```bash
+apiVersion: kaito.sh/v1alpha1
+kind: Workspace
+metadata:
+  name: workspace-tuning-phi-3
+resource:
+  instanceType: Standard_NC24ads_A100_v4
+  labelSelector:
+    matchLabels:
+      app: tuning-phi-3
+tuning:
+  preset:
+    name: phi-3-mini-128k-instruct
+  method: qlora
+  input:
+    urls:
+      - https://huggingface.co/datasets/myuser/mydataset/resolve/main/data/train-00000-of-00001.parquet?download=true
+  output:
+    image: myregistry.azurecr.io/adapters/myadapter:0.0.1
+    imagePushSecret: myregistrysecret
+```
+
+For the most part, the `Workspace` resource is similar as before. Here, we see that KAITO recommends a minimum GPU VM SKU of “Standard_NC24ads_A100_v4” for the tuning job.  You can update the `instanceType` to a larger GPU VM SKU, but either way, you should make sure there’s enough GPU quota in your Azure subscription prior to deploying applying the workspace to your AKS cluster. 
```

Comment:
```suggestion
The tuning property specifies the minimum information required for starting a tuning job, including:

**Preset**: indicates the target model for tuning (KAITO uses preset configurations to generate all Kubernetes sources needed to run the Job). 
**Method**: users can choose LoRA or QLoRA (quantized LoRA, for lower GPU memory usage) as the tuning method. 
**Input**: for flexibility, the tuning input can be a URL (for a public dataset), or images (for a private dataset).
**Output**: where the adapter is stored, as a container image or any other storage type supported by Kubernetes.

> The choice of GPU SKU is critical since model fine-tuning normally requires more GPU memory compared to model inference. To avoid GPU Out-Of-Memory, NVIDIA A100 or higher tier GPUs are recommended. 

You can check out the [KAITO tuning API](https://github.com/Azure/kaito/tree/main/docs/tuning) for more details and troubleshooting tips!
```